consensus/src/chained_bft/persistent_storage_test.rs

```diff
@@ -0,0 +1,271 @@
+    let (root_block, root_quorum_cert, root_ledger_info) =
+        RecoveryData::find_root(&mut blocks.clone(), &mut quorum_certs, &storage_ledger).unwrap();
+
+    // Verify that LI(S) is the root.
+    assert_eq!(*li_storage, root_block);
+    assert_eq!(*li_storage_child.quorum_cert(), root_quorum_cert);
+    assert_eq!(*li_storage_grandchild.quorum_cert(), root_ledger_info);
+}
+
+/// covers
+///  - blocks: simple tree
+///  - quorum_certs: simple tree
+///  - storage_ledger: disconnect
+///  - validity: root found
+///  - LI(S) and LI(C) relationship: LI(S) does not exist
+///
+/// ==> LI(C) is the root block
+#[test]
+fn test_tree_restartability_root_from_consensusdb() {
+    let (executed_blocks, _) = build_simple_tree();
+    let (mut blocks, mut quorum_certs) = get_find_root_params(executed_blocks.clone());
+
+    // Use a random hash value as the consensus block id.
+    // (i.e. LI(S) does not exist.)
+    let storage_ledger = placeholder_ledger_info_for_consensus_block_id(HashValue::random());
+
+    // Subset of blocks is inputted to `find_root()` so that LI(S) is disconnected to LI(C).
+    let (root_block, root_quorum_cert, root_ledger_info) =
+        RecoveryData::find_root(&mut blocks, &mut quorum_certs, &storage_ledger).unwrap();
+
+    // Unpack the longest valid chain in the tree. See `build_simple_tree()` diagram.
+    let a1 = executed_blocks[1].block();
+    let a2 = executed_blocks[2].block();
+    let a3 = executed_blocks[3].block();
+
+    // Verify that LI(C), or a1, is the root.
+    assert_eq!(*a1, root_block);
+    assert_eq!(*a2.quorum_cert(), root_quorum_cert);
+    assert_eq!(*a3.quorum_cert(), root_ledger_info);
+}
+
+/// covers
+///  - blocks: simple tree
+///  - quorum_certs: simple tree
+///  - storage_ledger: root
+///  - validity: root found
+///  - LI(S) and LI(C) relationship: LI(S) exist && LI(S) is an ancestor of LI(C)
+///                                  (note: a block is considered to be an ancestor of itself)
+///
+/// ==> LI(S) is the root block
+#[test]
+fn test_tree_restartability_same_root() {
+    let (executed_blocks, _) = build_simple_tree();
+    let (mut blocks, mut quorum_certs) = get_find_root_params(executed_blocks.clone());
+
+    // Construct the storage ledger using the id of a1.
+    let storage_ledger = placeholder_ledger_info_for_consensus_block_id(executed_blocks[1].block().id());
+
+    // Subset of blocks is inputted to `find_root()` so that LI(S) is disconnected to LI(C).
+    let (root_block, root_quorum_cert, root_ledger_info) =
+        RecoveryData::find_root(&mut blocks, &mut quorum_certs, &storage_ledger).unwrap();
+
+    // Unpack the longest valid chain in the tree. See `build_simple_tree()` diagram.
+    let a1 = executed_blocks[1].block();
+    let a2 = executed_blocks[2].block();
+    let a3 = executed_blocks[3].block();
+
+    // Verify that LI(S), or a1, is the root. (Note: LI(S) == LI(C))
+    assert_eq!(*a1, root_block);
```

Comment:
there're a lot boilerplate here, we can probably abstract it with a function like
```
fn expect_root(expected_root: &[Block<TestPayload], found_root: (Block, QC, QC)) -> bool {
  expected_root[0] == found_root.0 && ...
}
```

Reply:
I agree, but unfortunately, depending on the length of the chain or other commit situations, the relationship among the root, the quorum cert, and the ledger info may change. I've clarified this point in the code as well.

Reply:
isn't the expected_root always a 3-chain? I see it's either blocks[idx..idx+3] or executed_blocks[1..4]?

Reply:
Resolved offline.

- Typical scenario: `B0 <- [C0 <-B1] <- [C1 <- B2] <- [C2 <- B3] <- [C3 <- B4]` => `find_root()` => **(B1, C1, C3)**
- Genesis root scenario: `B0 <- [C0 <-B1] <- [C1 <- B2] <- [C2 <- B3]` (or shorter)  => `find_root()` => **(B0, C0, C0)** (i.e. C2 is not re-committed to ledger info)

Note: `blocks` chain must be of length 4 or greater except when genesis is the root. If genesis is the root, the `blocks` chain can be as small as length 1.